Add `ac_fixed` Sample (#1)
wangdi4/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md

The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.

| Optimized for                     | Description
---                                 |---
| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04* 
| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit

> | What you will learn               | How different methods of `ac_fixed` number construction affect hardware resource utilization <br> Recommended method for constructing `ac_fixed` numbers in your kernel <br> Accessing and using the `ac_fixed` math library functions <br> Trading off accuracy of results for reduced resource usage on the FPGA

Author: trade off  /  trade-off  /  tradeoff
trading off  /  trading-off

| Time to complete                  | 30 minutes

## Purpose

This FPGA tutorial shows how to use the `ac_fixed` type with some simple examples.

This data type can be used in place of native floating point types to generate area efficient and optimized designs for the FPGA. Operations which do not utilize the full dynamic range of the native types are good candidates for using `ac_fixed` types. For example: multiplying by a number in the range of `(-1.0, 1.0)`.

This tutorial shows the recommended method for constructing an `ac_fixed` number, some examples of using the fixed point math library functions and how they can be used to reduce the area of the hardware generated by the compiler by trading-off accuracy of the mathematical operations.

### Simple Code Example

An `ac_fixed` number can be defined as follows:
```cpp
ac_fixed<W, I, S> a;
```
Here `W` specifies the width and `S` specifies the sign of the number. One of the `W` bits is used to store the sign information. The second parameter `I` is an integer that specifies the location of the fixed point relative to the most significant bit.

> The type also provides two more optional parameters for controlling the overflow and rounding modes. For more details on the type, please refer to the section titled `Variable-Precision Integer and Floating-Point Support` in the IntelÂ® oneAPI DPC++ FPGA Optimization Guide.

Author: Intel oneAPI ... Guide hyperlink?

To use an `ac_fixed` type in your code, you must include the following header:

```cpp
#include <sycl/ext/intel/ac_types/ac_fixed.hpp>
```
Additionally, you must pass the flag `-qactypes` on Linux or `/Qactypes` on Windows to the `dpcpp` command when compiling your SYCL program in order to ensure that the headers are correctly included. In this tutorial, this is done in `src/CMakeLists.txt`.

### Recommended Method for Constructing `ac_fixed` Numbers

The compiler uses significant FPGA resources to convert double precision (and single precision) floating-point values to `ac_fixed` values. The kernel `ConstructFromFloat` constructs an `ac_fixed` object from an accessor to a native `float` type.

> In contrast, the kernel `ConstructFromACFixed` constructs an `ac_fixed` object from an accessor to another `ac_fixed` object. This consumes far less area than the previous kernel. See the section on examining the reports below to understand where to look for this difference within the optimization reports.

Author: "Examining the Report"

### Using the `ac_fixed` Math Functions

> To use this type in your code, you must include the following header:

Author: To use `ac_fixed` math functions, ...

```cpp
#include <sycl/ext/intel/ac_types/ac_fixed_math.hpp>
```

> The flag `-qactypes` on Linux or `/Qactypes` on Windows  will ensure that the compiler includes the header and links against the necessary libraries for emulation of the math library functions.

Author: Additionally, you must pass the flag `-qactypes` on Linux or `/Qactypes` on Windows to the `dpcpp` command when compiling your SYCL program ~~in order to ensure that the headers are correctly included~~. In this tutorial, this is done in `src/CMakeLists.txt`.
=>
...for emulation of the math library functions...

> This tutorial design contains two kernels `CalculateWithFloat` and `CalculateWithACFixed`. Both calculate the simple expression:

Author: contains two kernels that use the `ac_fixed` math functions: `CalculateWithFloat` and `CalculateWithACFixed`. ...

```cpp
  square_root ( sine(x) * sine(x) + cosine(x) * cosine(x) )
```
for some input `x`.

The kernel `CalculateWithFloat` uses floating point values and the standard math library while `CalculateWithACFixed` uses `ac_fixed` values and the `ac_fixed` math library. The `ac_fixed` inputs are instantiated with the following parameters:

```cpp
  W = 10, I = 3, S = true
```

> Clearly, the `ac_fixed` numbers are smaller in size than floating point numbers. This results in reduction of the FPGA resources utilized by the functions at the expense of accuracy. To see the trade-offs between accuracy compare the numeric results of the operations. The area utilization differences will be discussed in the section on `Examining the reports`.

Author: ```suggestion
Clearly, the `ac_fixed` numbers are smaller in size than floating point numbers. This results in reduction of the FPGA resources utilized by the functions at the expense of accuracy. To see the trade-offs between accuracy, compare the numeric results of the operations. The area utilization differences will be discussed in the section on `Examining the reports`.
```

## License

Code samples are licensed under the MIT license. See
[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.

## Building the `ac_fixed` Tutorial

### Include Files

The included header `dpc_common.hpp` is located at `%ONEAPI_ROOT%\dev-utilities\latest\include` on your development system.

### Running Samples in DevCloud
If running a sample in the Intel DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the Intel® oneAPI Base Toolkit Get Started Guide ([https://devcloud.intel.com/oneapi/documentation/base-toolkit/](https://devcloud.intel.com/oneapi/documentation/base-toolkit/)).

When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.

> 

Author: Copy `Using Visual Studio Code*  (Optional)` section from ac_int tutorial

* **FPGA hardware compile**: `ac_fixed.prj`

Scroll down on the Summary page of the report and expand the section titled `Compile Estimated Kernel Resource Utilization Summary`. Observe how the kernel `ConstructFromACFixed` consumes lesser resources than the kernel named `ConstructFromFloat`. Similarly, observe how the kernel named `CaclulateWithACFixed` consumes lesser FPGA resources than `CalculateWithFloat`.

## Running the Sample

1. Run the sample on the FPGA emulator (the kernel executes on the CPU):

   ```bash
   ./ac_fixed.fpga_emu    # Linux
   ac_fixed.fpga_emu.exe  # Windows
   ```

2. Run the sample on the FPGA device

   ```bash
   ./ac_fixed.fpga             # Linux
   ```

### Example of Output

```txt
Constructed from float:         3.6416015625
Constructed from ac_fixed:      3.6416015625

MAX DIFF for ac_fixed<10, 3, true>:  0.0078125
MAX DIFF for float:                  9.53674e-07

result(fixed point): 1
result(float):       1

result(fixed point): 0.992188
result(float):       1

result(fixed point): 1
result(float):       1

result(fixed point): 1
result(float):       1

result(fixed point): 0.992188
result(float):       1

> PASSED

Author: PASSED: ...

| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
| What you will learn               | How different methods of `ac_fixed` number construction affect hardware resource utilization <br> Recommended method for constructing `ac_fixed` numbers in your kernel <br> Accessing and using the `ac_fixed` math library functions <br> Trading off accuracy of results for reduced resource usage on the FPGA
| Time to complete                  | 30 minutes

## Purpose

This FPGA tutorial shows how to use the `ac_fixed` type with some simple examples.

This data type can be used in place of native floating point types to generate area efficient and optimized designs for the FPGA. Operations which do not utilize the full dynamic range of the native types are good candidates for using `ac_fixed` types. For example: multiplying by a number in the range of `(-1.0, 1.0)`.

> This tutorial shows the recommended method for constructing an `ac_fixed` number, some examples of using the fixed point math library functions and how they can be used to reduce the area of the hardware generated by the compiler by trading off accuracy of the mathematical operations.

Reviewer: ```suggestion
This tutorial shows the recommended method for constructing an `ac_fixed` number, some examples of using the fixed point math library functions, and how they can be used to reduce the area of the hardware generated by the compiler by trading off accuracy of the mathematical operations.
```

### Simple Code Example

An `ac_fixed` number can be defined as follows:
```cpp
ac_fixed<W, I, S> a;
```

> Here `W` specifies the width and `S` specifies the sign of the number. One of the `W` bits is used to store the sign information. The second parameter `I` is an integer that specifies the location of the fixed point relative to the most significant bit.

Reviewer: Maybe a little ASCII art with an example? Like:
```For example, ac_int<10, 4, true> would give you the following <insert a little ASCII art about the bits and their positions".
```

Also, does S = true add a bit? Or is the decimal point 'I' positions from the sign bit / the MSB? Again, I think a few examples with ASCII art may help here.

Author: Mention that the sign bit is the MSB of the W bits, and direct users to `ac_datatypes_ref.pdf` for more details.

> The type also provides two more optional parameters for controlling the overflow and rounding modes. For more details on the type, please refer to the section *Variable-Precision Integer and Floating-Point Support* in the Intel® oneAPI DPC++ FPGA Optimization Guide.

Reviewer: ```suggestion
The `ac_fixed` class also provides two more optional parameters for controlling the overflow and rounding modes. For more details on the `ac_int` class, refer to the section *Variable-Precision Integer and Floating-Point Support* in the Intel® oneAPI DPC++ FPGA Optimization Guide.
```

To use an `ac_fixed` type in your code, you must include the following header:

```cpp
#include <sycl/ext/intel/ac_types/ac_fixed.hpp>
```
Additionally, you must pass the flag `-qactypes` on Linux or `/Qactypes` on Windows to the `dpcpp` command when compiling your SYCL program in order to ensure that the headers are correctly included. In this tutorial, this is done in `src/CMakeLists.txt`.

> ### Recommended Method for Constructing `ac_fixed` Numbers

Reviewer: Do we have the same gotchas here that we must initialize the ac_fixed to 0 if we want to manipulate individual bits? If so, it may be worth mentioning. (https://hsdes.intel.com/appstore/article/#/14015829976)

In contrast, the kernel `ConstructFromACFixed` in the function `TestConstructFromACFixed` constructs an `ac_fixed` object from an accessor to another `ac_fixed` object. This consumes far less area than the previous kernel. See the section on *Examining the Reports* below to understand where to look for this difference within the optimization reports.

### Using the `ac_fixed` Math Functions

To use the `ac_fixed` math functions in your code, you must include the following header:

```cpp
#include <sycl/ext/intel/ac_types/ac_fixed_math.hpp>
```

The functions `TestCalculateWithFloat` and `TestCalculateWithACFixed` in this tutorial design contain the kernels `CalculateWithFloat` and `CalculateWithACFixed` respectively. Both calculate the simple expression:
```cpp
  square_root ( sine(x) * sine(x) + cosine(x) * cosine(x) )
```
for some input `x`.

The kernel `CalculateWithFloat` uses floating point values and the standard math library while `CalculateWithACFixed` uses `ac_fixed` values and the `ac_fixed` math library. The `ac_fixed` inputs are instantiated with the following parameters:

```cpp
  W = 10, I = 3, S = true
```

> Clearly, the `ac_fixed` numbers are smaller in size than floating point numbers. This results in reduction of the FPGA resources utilized by the functions at the expense of accuracy. To see the trade-offs between accuracy compare the numeric results of the operations. The area utilization differences will be discussed in the section on *Examining the Reports*.

Reviewer: ```suggestion
The `ac_fixed` numbers are smaller in size than floating point numbers, which results in a reduction of the FPGA resources at the expense of accuracy. To see the trade-offs between accuracy compare the numeric results of the operations. The area utilization differences will be discussed in the section on *Examining the Reports*.
```

Additionally, I don't like this part "To see the trade-offs between accuracy compare the numeric results of the operations". Do we do that for the user? We should show the accuracy differences in the tutorial.

When you use the `ac_fixed` library, keep the following points in mind:

1. Input Bit Width and Input Value Range Limits

    The fixed-point math functions have bit width and input value range requirements. All bit width and input value range requirements are documented at the top of the `ac_fixed_math.hpp` file. For example, the `sin_fixed` and `cos_fixed` functions require the integer part's bit width to be 3, and the input value range to be within [-pi, pi].

2. Return Types

    For fixed-point functions, each function has a default return type. Assigning the result to a non-default return type triggers a type conversion and can cause an increase in logic use or a loss of accuracy in your results.  All return types are documented at the top of the `ac_fixed_math.hpp` file. For example, for `sin_fixed` and `cos_fixed`, the input type is `ac_fixed<W, 3, true>`, and the output type is `ac_fixed<W-1, 2, true>`.

3. Accuracy
 - Floating point vs Fixed point

>     The host program (`main()` function) for this tutorial gives you an estimate of the difference between the correct result and the result provided by the math library functions. The floating point version (which has a greater bit width in this case) generates a more accurate result.

Reviewer: ```suggestion
    The host program for this tutorial shows the accuracy differences between the correct result and the result provided by the math library functions. The floating point version generates a more accurate result than the fixed point version.
```

1) Nevermind my earlier comment about showing this :)
2) What is the "correct" result? Do you list out EVERY decimal?

>  - Emulation vs Simulation for fixed point math operations

Reviewer: Can we say emulation vs hardware here? Simulation support for these designs is not ready yet.

>     Due to the differences in the internal math implementations, the results from `ac_fixed` math functions in simulation and emulation might not always be bit-accurate. In this example you can observe the difference between emulation and simulation.

Reviewer: I am not sure we can claim support for simulation yet. Do these differences show up between emulation and hardware too? If so, show that instead.

Reviewer: In fact, your cmake file doesn't have a simulation target, so certainly change this.

This tutorial shows the recommended method for constructing an `ac_fixed` number, some examples of using the fixed point math library functions, and how they can be used to reduce the area of the hardware generated by the compiler by trading off accuracy of the mathematical operations.

### Simple Code Example

An `ac_fixed` number can be defined as follows:
```cpp
ac_fixed<W, I, S> a;
```
Here `W` specifies the width in bits and `S` is a bool indicating if the number is signed. Signed numbers use the most significant bit (MSB) of the `W` bits to store the sign bit. The second parameter `I` is an integer that specifies the location of the fixed point relative to the MSB.

The `ac_fixed` class also provides two more optional parameters for controlling the overflow and rounding modes. 

For more details on the `ac_fixed` class, refer to the documentation provided with the installation at `%INSTALL_DIR%/include/sycl/ext/intel/ac_types/ac_datatypes_ref.pdf`.

To use an `ac_fixed` type in your code, you must include the following header:

```cpp

> #include <sycl/ext/intel/ac_types/ac_fixed.hpp>

Reviewer 2: `./compiler/2022.0.2/linux/include/sycl/CL/sycl/INTEL/ac_types/ac_fixed.hpp`
and
`./compiler/2022.0.2/linux/lib/oclfpga/include/CL/sycl/INTEL/ac_types/ac_fixed.hpp`
- We keep this for backwards compatibility. We don't need two of these. I will follow up with others to figure why there are two of these.

`./compiler/2022.0.2/linux/lib/oclfpga/include/sycl/ext/intel/ac_types/ac_fixed.hpp` - This is the current location.

```cpp
  square_root ( sine(x) * sine(x) + cosine(x) * cosine(x) )
```
for some input `x`.

The kernel `CalculateWithFloat` uses floating point values and the standard math library while `CalculateWithACFixed` uses `ac_fixed` values and the `ac_fixed` math library. The `ac_fixed` inputs are instantiated with the following parameters:

```cpp
  W = 10, I = 3, S = true
```

The `ac_fixed` numbers are smaller in size than floating point numbers, which results in a reduction of the FPGA resources at the expense of accuracy. To see the trade-offs between accuracy compare the numeric results of the operations. The area utilization differences will be discussed in the section on *Examining the Reports*.

When you use the `ac_fixed` library, keep the following points in mind:

1. Input Bit Width and Input Value Range Limits

    The fixed-point math functions have bit width and input value range requirements. All bit width and input value range requirements are documented at the top of the `ac_fixed_math.hpp` file. For example, the `sin_fixed` and `cos_fixed` functions require the integer part's bit width to be 3, and the input value range to be within [-pi, pi].

2. Return Types

>     For fixed-point functions, each function has a default return type. Assigning the result to a non-default return type triggers a type conversion and can cause an increase in logic use or a loss of accuracy in your results.  All return types are documented at the top of the `ac_fixed_math.hpp` file. For example, for `sin_fixed` and `cos_fixed`, the input type is `ac_fixed<W, 3, true>`, and the output type is `ac_fixed<W-1, 2, true>`.

Reviewer 3: you can also use the `auto` type to let the compiler use the default return type. 
bonus: mousing over the `auto` type in VS Code with C++ extension will show you what the type will evaluate to.

Author: Is it a better convention to declare the `ac_fixed` type explicitly, so users would be aware if loss of accuracy happens?

```cpp
ac_fixed<W, I, S> a;
```
Here `W` specifies the width in bits and `S` is a bool indicating if the number is signed. Signed numbers use the most significant bit (MSB) of the `W` bits to store the sign bit. The second parameter `I` is an integer that specifies the location of the fixed point relative to the MSB. Here are some examples of the range and quantum of `ac_fixed` numbers.

```cpp
ac_fixed<4,4,true> x; // xxxx. : range = [-8, 7], quantum = 1
ac_fixed<4,0,true> x; // .xxxx : range = [-8/16, 7/16], quantum = 1/16
ac_fixed<4,0,false> x; // .xxxx : range = [0, 15/16], quantum = 1/16
ac_fixed<4,7,false> x; // xxxx000. : range = [0, 120], quantum = 8
ac_fixed<4,-3,true> x; // (.xxxx)2^(-3) : range=[-1/16, 7/128], quantum = 1/128
```

The `ac_fixed` class also provides two more optional parameters for controlling the overflow and rounding modes. 

> For more details on the `ac_fixed` class, please refer to the documentation provided with the installation at `<installation path>/lib/oclfpga/include/ref/ac_datatypes_ref.pdf`.

Reviewer 3: sorry to change this again, here is an external link that points to the correct version of the doc: https://github.com/hlslibs/ac_types/blob/v3.7/pdfdocs/ac_datatypes_ref.pdf

Reviewer 2: Hi @paul-white-intel , is there a concern about the new updated path we have put here? I don't think we want to point to an external doc URL which may change without notice.

Reviewer 3: we will eventually host our own copy of the ac_int docs; Eric and VJ are supposed to be working on that (we got endorsement from steering committee last week). My thought was that we would point to the github link until we have an intel.com/.../ac_datatypes_ref.pdf link available; but if you would like to avoid thrashing on the code samples we can link to the [oneAPI FPGA optimization guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/optimize-your-design/resource-use/data-types-and-operations/var-prec-fp-sup.html), and link to the Siemens doc from our optimization guide.

Reviewer 4: Intel-hosted version is now available at https://cdrdv2.intel.com/v1/dl/getContent/728986
